### ISSUE TITLE
Uploaded avatars moved before cropping

### ISSUE THREAD

**Issue by fjelltid** [humanmade/S3-Uploads#94]
Hi, 

Thank you for an awesome plugin. :)
It works almost perfect for us. We are running it on a site with WP and BP.
The issue we have is when copping an image after upload. The avatar get uploaded locally just fine, but if you need to crop it, the trouble begin. It looks like the local file is removed before the cropping and saving is done. The file does not get saved and not moved to S3 of these reason.

I have tried to do some debugging (I´m not a programmer, but I´ll do my best) and found the following errors (and have attached the lines from the mentioned file) when uploading and trying to crop:

PHP Warning:  closedir() expects parameter 1 to be resource, boolean given in /var/app/current/wp-content/plugins/buddypress/bp-core/bp-core-avatars.php on line 524

/var/app/current/wp-content/plugins/buddypress/bp-core/bp-core-avatars.php:
523:		// Close the avatar directory.
524:		closedir( $av_dir );
525:
526:		// If we found a locally uploaded avatar.
527:		if ( isset( $avatar_url ) ) {
528:			// Support custom scheme.
529:			$avatar_url = set_url_scheme( $avatar_url, $params['scheme'] );
530:
531:			// Return it wrapped in an <img> element.
532:			if ( true === $params['html'] ) {

PHP Warning:  opendir(/var/app/current/wp-content/uploads/avatars/6277): failed to open dir: Permission denied in /var/app/current/wp-content/plugins/buddypress/bp-core/bp-core-avatars.php on line 482
/var/app/current/wp-content/plugins/buddypress/bp-core/bp-core-avatars.php:
469:	/**
470:	 * Look for uploaded avatar first. Use it if it exists.
471:	 * Set the file names to search for, to select the full size
472:	 * or thumbnail image.
473:	 */
474:	$avatar_size              = ( 'full' == $params['type'] ) ? '-bpfull' : '-bpthumb';
475:	$legacy_user_avatar_name  = ( 'full' == $params['type'] ) ? '-avatar2' : '-avatar1';
476:	$legacy_group_avatar_name = ( 'full' == $params['type'] ) ? '-groupavatar-full' : '-groupavatar-thumb';
477:
478:	// Check for directory.
479:	if ( file_exists( $avatar_folder_dir ) ) {
480:
481:	// Open directory.
482:	if ( $av_dir = opendir( $avatar_folder_dir ) ) {
483:
484:		// Stash files in an array once to check for one that matches.
485:		$avatar_files = array();
486:		while ( false !== ( $avatar_file = readdir( $av_dir ) ) ) {
487:			// Only add files to the array (skip directories).
488:			if ( 2 < strlen( $avatar_file ) ) {
489:				$avatar_files[] = $avatar_file;
490:			}
491:		}

PHP Fatal error:  Cannot use object of type WP_Error as array in /var/app/current/wp-content/plugins/buddypress/bp-core/bp-core-avatars.php on line 908, referer: https://www.xxxxxxx.xxx/<username>/profile/change-avatar/
/var/app/current/wp-content/plugins/buddypress/bp-core/bp-core-avatars.php:
903:// We only want to handle one image after resize.
904:	if ( empty( $bp->avatar_admin->resized ) ) {
905:		$bp->avatar_admin->image->file = $bp->avatar_admin->original['file'];
906:		$bp->avatar_admin->image->dir  = str_replace( $upload_path, '', $bp->avatar_admin->original['file'] );
907:	} else {
908:		$bp->avatar_admin->image->file = $bp->avatar_admin->resized['path'];
909:		$bp->avatar_admin->image->dir  = str_replace( $upload_path, '', $bp->avatar_admin->resized['path'] );
910:		@unlink( $bp->avatar_admin->original['file'] );
911:	}
 
I don´t know if this is of any help to resolve it, but i really hope it can help a bit. 
If you need an server to test it on, I can provide this via e-mail.

Thank you so much.

**Comment by fjelltid** [comment]
@joehoyle - Any ideas, suggestions or hits to get me closer to an solution?
 I´m really sorry for bothering you with this, but I don´t know where else to turn.

Thank you once again for an awesome plugin. :)

**Comment by joehoyle** [comment]
Hmm this looks as if this part of BP is not handling custom stream wrappers very well, however I don't really understand why it would be implementing it's own streaming rather than using the WordPress APIs like wp_handle_upload / download_url / handle_upload_bits, maybe @paulgibbs could elaborate on that as I seem to remember him working on the avatars code for WordPress.com VIP or something alike that.

**Comment by paulgibbs** [comment]
tl;dr: I don't have any immediate suggestions, other than to dive into the code and start seeing exactly what's breaking, and then workaround it it. :(

  

We do use WordPress functions for uploading images, fixing rotations, resizing, etc.

Lots of avatar code in BuddyPress is 8 year old code, which needs re-architecturing but that's a very tricky job and no-one has gotten to yet. :) @imath tidied up some of the internal APIs about a year ago, but there's a lot to be done -- the work I did for Automattic VIP proved it.

The plugin that supports BuddyPress for Automattic VIP is here https://github.com/humanmade/buddypress-vip-go but Automattic use some kind of remote service on top of MogileFS, so that plugin would not be a drop-in solution for @fjelltid's problem, though the approach take may be of use to other developers.

We do save the uploaded image to the filesystem before (optionally) cropping it afterwards. If S3 Uploads is doing something like uploading the image immediately and removing the local version, I could imagine maybe the `$bp->avatar_admin->original` global would now have the wrong path to the image. I suspect that'd be the first issue, but there are likely more.

The weirdest part IMO is that we have a weird system to find avatars for a piece of content by using a scandir technique and fetching the first image found, which is what some of the code in the ticket here is handling -- which could well be causing problems with external file systems. I had to do a lot of filtering and bypassing and duplication to get it to work nicely on Automattic VIP.

**Comment by fjelltid** [comment]
Sorry for the late replay.
Thank you so much both @joehoyle and @paulgibbs for your answers and help.

I will try to dive into the code and see what I can find. If I don´t manage to find or solve the issue, where do you recommend me to get or hire help to fix this?

**Comment by joehoyle** [comment]
Closing due to inactive.
